### PR TITLE
Manage user Data

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
- #35 extract all user data from Gui and into _UserData
- #37 add a render property to chart
- #6 assign dictionnary

### CODE REVIEW

File: taipy/gui/utils/_user_data.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import typing as t
+
+from .MapDictionary import _MapDictionary
+from ..data.data_scope import _DataScopes
+
+if t.TYPE_CHECKING:
+    from ..gui import Gui
+
+
+class _UserData:
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_user_data.py]: _UserData is far too vague for me.
What do you think about `_Bindings`?

File: taipy/gui/utils/__init__.py [hunk 2]
```diff
@@ -21,3 +21,4 @@
     TaipyContent,
     TaipyContentImage,
 )
+from .attributes import delscopeattr, getscopeattr, getscopeattr_drill, getuserattr, hasscopeattr, hasuserattr, setscopeattr, setscopeattr_drill
```
Comment (FabienLelaquais) [inline taipy/gui/utils/__init__.py]: Do we really need those?

File: taipy/gui/utils/_state.py [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
+
+if t.TYPE_CHECKING:
+    from ..gui import Gui
+
 
 class State:
     __attrs = ("_gui", "_user_var_list")
 
-    def __init__(self, gui: t.Any) -> None:
+    def __init__(self, gui: "Gui") -> None:
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_state.py]: No type for 'gui', and what about this default value?!

File: taipy/gui/utils/attributes.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from operator import attrgetter
```
Comment (FabienLelaquais) [inline taipy/gui/utils/attributes.py]: I suspect this file would be '_attributes.py'?

File: taipy/gui/utils/attributes.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    return getattr(gui._get_user_data(), name)
+
+
+def hasuserattr(gui: "Gui", name: str) -> bool:
+    return hasattr(gui._get_user_data(), name)
+
+
+def getscopeattr(gui: "Gui", name: str, *more) -> t.Any:
+    if more:
+        return getattr(gui._get_user_data()._get_data_scope(), name, more[0])
```
Comment (FabienLelaquais) [inline taipy/gui/utils/attributes.py]: Why 'more[0]' and not 'more'? I guess that would be more useful...

File: tests/taipy/gui/gui_specific/test_mapdict.py [hunk 6]
```diff
@@ -91,4 +95,13 @@ def update(k, v):
...
\ No newline at end of file
+    assert temp_values["b"] == 5
+
+
+def test_mapdict_set(gui: Gui):
+    d = {"a": 1}  # noqa: F841
+    gui.run(run_server=False)
+    assert isinstance(gui._Gui__state.d, _MapDictionary)
+    gui._Gui__state.d = {"b": 2}
+    assert isinstance(gui._Gui__state.d, _MapDictionary)
```
Comment (FabienLelaquais) [inline tests/taipy/gui/gui_specific/test_mapdict.py]: No checks on the values?

File: taipy/gui/gui.py [hunk 7]
```diff
@@ -152,10 +159,8 @@ def __get_content_accessor(self):
             self.__content_accessor = ContentAccessor(self._get_app_config("data_url_max_size", 50 * 1024))
         return self.__content_accessor
 
-    def __get_state(self):
-        if self.__state is None:
-            self.__state = State(self)
-        return self.__state
+    def _get_user_data(self):
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: ... and that would be '_bindings()'... (see other remark on naming)

File: taipy/gui/gui.py [hunk 8]
```diff
@@ -723,24 +686,28 @@ def add_partial(
 
     # Main binding method (bind in markdown declaration)
     def bind_var(self, var_name: str) -> bool:
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py:689]: I think we could group `bind_func()` and `bind_var()` in the same `bind()` function.
I wonder if `bind_var_val()` is of any use...

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Minor comment but including renaming, potentially.
You have to talk to me abot *drill.
Looks really nice though

File: taipy/gui/gui.py (hunk 8 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py:689]: in theory bind_func doesn't need the scope ...

File: taipy/gui/gui.py (hunk 7 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: OK

File: tests/taipy/gui/gui_specific/test_mapdict.py (hunk 6 above)
Comment (FredLL-Avaiga) [reply tests/taipy/gui/gui_specific/test_mapdict.py]: I could ...

File: taipy/gui/utils/attributes.py (hunk 5 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/attributes.py]: I think more is a list (or Iterable) ...

File: taipy/gui/utils/attributes.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/attributes.py]: ok

File: taipy/gui/utils/_state.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_state.py]: ?

File: taipy/gui/utils/__init__.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/__init__.py]: we do

File: taipy/gui/utils/_user_data.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_user_data.py]: Agreed

File: taipy/gui/utils/attributes.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+from operator import attrgetter
+import typing as t
+
+from . import attrsetter
+
+if t.TYPE_CHECKING:
+    from ..gui import Gui
+
+
+def getuserattr(gui: "Gui", name: str, *more) -> t.Any:
```
Comment (dinhlongviolin1) [inline taipy/gui/utils/attributes.py]: I have never seen this, is it ok for Gui to be in double quotes

File: taipy/gui/utils/attributes.py (hunk 4 above)
Comment (dinhlongviolin1) [reply taipy/gui/utils/attributes.py]: Nice utils file

**Review by dinhlongviolin1 — APPROVED** [review]
A good improvement to the structure. I must admit I don't like the variable binding to Gui itself.  Binding user variable to user data is a much better way of encapsulating data binding. Great job Fred!

File: taipy/gui/utils/_state.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_state.py]: it's a way to work around circular dependencies

File: taipy/gui/utils/attributes.py (hunk 9 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/attributes.py]: for circular dependencies ...

File: taipy/gui/gui.py (hunk 8 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py:689]: bind_var_val : 115 occurrences in 25 files
OK mainly in tests
but outside of tests: 7 occurrences in 3 files

**Review by FabienLelaquais — APPROVED** [review]
Go for it!

**Pull request merged**